Sources/MongoSwift/SDAM.swift

```diff
@@ -231,29 +254,137 @@ public struct TopologyDescription {
         }
     }
 
+    /// An internal function used to calculate staleness of a given server in a replica set
+    internal func staleness(for server: ServerDescription) -> Double? {
+        guard [.replicaSetWithPrimary, .replicaSetNoPrimary].contains(self.type) else {
+            return nil
+        }
+
+        // From spec: Non-secondary servers have zero staleness
+        guard server.type == .rsSecondary else {
+            return 0.0
+        }
+
+        // Can force unwrap lastWriteDate in here since it's always present in the the ismaster output of primaries
+        // and secondaries, and we filter out all other types of servers.
+        if self.type == .replicaSetWithPrimary {
+            let primary = self.servers.filter { server in server.type == .rsPrimary }[0]
+
+            let delta = { (serverDesc: ServerDescription) -> Double? in
+                guard let serverLastUpdate = serverDesc.lastUpdateTime else {
+                    return nil
+                }
+                return (serverLastUpdate.timeIntervalSinceReferenceDate -
+                        serverDesc.lastWriteDate!.timeIntervalSinceReferenceDate) * 1000.0
+            }
+
+            guard let clientToServer = delta(server), let clientToPrimary = delta(primary) else {
+                return nil
+            }
+
+            return clientToServer - clientToPrimary + Double(TopologyDescription.heartbeatFrequencyMS)
+        } else {
+            let possibleSMax = self.servers.filter { $0.type == .rsSecondary }.max {
+                $0.lastWriteDate!.timeIntervalSinceReferenceDate < $1.lastWriteDate!.timeIntervalSinceReferenceDate
+            }
+
+            guard let sMax = possibleSMax else {
```

Comment:
so this happens if there ended up being no secondaries, right? can we check for that before we call max? otherwise it's a little hard to parse what edge case this is checking for.

Reply:
Yep, re-ordered for clarity.